### DISCUSSION TITLE
Error: the trait `tokio::io::async_read::AsyncRead` is not implemented for `tokio::net::TcpStream`

### DISCUSSION THREAD

**Discussion by Urhengulas** [tokio-rs/tokio#2984] [Q&A] [answered]
Though `AsyncRead` and `AsyncWrite` should be implemented for `TcpStream` (see [here](https://docs.rs/tokio/0.3.0/tokio/net/struct.TcpStream.html#trait-implementations)) rust tells me that it's not:

```
error[E0277]: the trait bound `tokio::net::TcpStream: tokio::io::async_read::AsyncRead` is not satisfied
   --> tarpc/src/serde_transport.rs:176:55
    |
176 |             Poll::Ready(Ok(new(self.config.new_framed(io), (self.codec_fn)())))
    |                                                       ^^ the trait `tokio::io::async_read::AsyncRead` is not implemented for `tokio::net::TcpStream`

```
Corresponding code: https://github.com/google/tarpc/blob/d27f341bdebe29df927b6d2b6f1a26fd657a0efe/tarpc/src/serde_transport.rs#L176

---

I also tried to enable the `"net"`-feature, but this actually changes nothing...

**Comment by Darksonn** [comment]
I do not get that error when I compile your project at that commit with `--all-features`.

**Reply by Urhengulas** [reply]
In the code I linked `tokio` isn't bumped to `0.3` yet.

But I think I resolved this by next to upgrading `tokio` to `0.3` also upgrading `tokio-util` to `0.4`..

**Comment by taiki-e** [comment, marked as answer]
@Urhengulas: You are probably using an old tokio-util. Can you check that tokio-util is 0.4?

**Reply by Urhengulas** [reply]
@taiki-e This solves the problem!

But I actually don't understand why... Could you please briefly explain this?

**Reply by Darksonn** [reply]
The `TcpStream` in Tokio 0.3 implements the `AsyncRead` trait from Tokio 0.3, but tokio-util 0.3 requires the `AsyncRead` trait from Tokio 0.2.
